File: backend/app/models/warehouse.py
"""
Warehouse Pydantic models — requests and responses for Phase 3.

Covers:
- Movement wizard (validate, preview, execute)
- Dashboard KPIs and activity
- Inventory grid
- Staging area
- Audit sessions and items
- Supplier preferences
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# =================================================================
# MOVEMENT WIZARD
# =================================================================

class MovementLineItem(BaseModel):
    """A single part+qty line within a movement request."""
    part_id: int
    qty: int = Field(ge=1)
    supplier_id: int | None = None  # resolved automatically if not provided


class MovementRequest(BaseModel):
    """Request to validate, preview, or execute a stock movement."""
    from_location_type: str
    from_location_id: int = 1
    to_location_type: str
    to_location_id: int = 1
    items: list[MovementLineItem] = Field(min_length=1, max_length=20)
    # Optional metadata
    reason: str | None = None
    reason_detail: str | None = None
    notes: str | None = None
    reference_number: str | None = None
    job_id: int | None = None
    photo_path: str | None = None
    scan_confirmed: bool = False
    gps_lat: float | None = None
    gps_lng: float | None = None
    # Staging destination hint (only for moves TO pulled)
    destination_type: str | None = None  # 'truck' or 'job'
    destination_id: int | None = None
    destination_label: str | None = None


class ValidationError(BaseModel):
    """A single validation error from pre-flight check."""
    field: str | None = None
    message: str
    part_id: int | None = None


class ValidationResult(BaseModel):
    """Result of movement validation."""
    valid: bool
    errors: list[ValidationError] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class MovementPreviewLine(BaseModel):
    """Preview of a single line item's before/after state."""
    part_id: int
    part_name: str
    part_code: str | None = None
    qty: int
    supplier_id: int | None = None
    supplier_name: str | None = None
    supplier_source: str | None = None  # 'preferred' or 'fifo'
    source_before: int
    source_after: int
    dest_before: int
    dest_after: int
    unit_cost: float | None = None
    line_value: float | None = None


class MovementPreview(BaseModel):
    """Full preview of a movement batch before execution."""
    lines: list[MovementPreviewLine]
    total_qty: int = 0
    total_value: float | None = None
    movement_type: str  # transfer, consume, return
    photo_required: bool = False
    warnings: list[str] = Field(default_factory=list)


class MovementResult(BaseModel):
    """Result of a single executed movement line."""
    movement_id: int
    part_id: int
    part_name: str
    qty: int
    movement_type: str
    from_location_type: str | None = None
    to_location_type: str | None = None


class MovementExecuteResponse(BaseModel):
    """Response from executing a batch movement."""
    success: bool = True
    movements: list[MovementResult] = Field(default_factory=list)
    total_items: int = 0
    total_qty: int = 0


# =================================================================
# MOVEMENT RULES
# =================================================================

MOVEMENT_RULES: dict[tuple[str, str], dict[str, Any]] = {
    ("warehouse", "pulled"):   {"type": "transfer", "photo_required": False},
    ("pulled", "truck"):       {"type": "transfer", "photo_required": False},
    ("warehouse", "truck"):    {"type": "transfer", "photo_required": False},
    ("truck", "job"):          {"type": "consume",  "photo_required": True},
    ("job", "truck"):          {"type": "return",   "photo_required": True},
    ("truck", "warehouse"):    {"type": "return",   "photo_required": False},
    ("pulled", "warehouse"):   {"type": "return",   "photo_required": False},
}

VALID_LOCATION_TYPES = {"warehouse", "pulled", "truck", "job"}

# Reason categories and sub-reasons for the wizard
REASON_CATEGORIES: dict[str, list[str]] = {
    "Job Stock": ["New Install", "Repair/Replace", "Service Call", "Warranty Work"],
    "Truck Restock": ["Daily Restock", "Special Order", "Emergency Load"],
    "Return": ["Unused/Leftover", "Wrong Part", "Job Cancelled", "Overstock"],
    "Damage/Loss": ["Damaged Transit", "Damaged on Job", "Lost/Missing", "Defective"],
    "Audit Adjustment": ["Count Correction", "Found Extra", "Write-off"],
    "Other": [],
}


# =================================================================
# DASHBOARD
# =================================================================

class DashboardKPIs(BaseModel):
    """Warehouse dashboard headline numbers."""
    stock_health_pct: float = 0.0       # % of parts within min/max range
    total_units: int = 0                 # total qty across warehouse
    warehouse_value: float | None = None  # $ value (permission-gated)
    shortfall_count: int = 0             # parts below min (excl. winding down)
    pending_task_count: int = 0          # staged items + audits + spot-checks


class ActivitySummary(BaseModel):
    """One-line summary of a recent movement for the activity feed."""
    id: int
    summary: str           # e.g. "Mike pulled 12× GFI White to Staging"
    movement_type: str
    performer_name: str | None = None
    created_at: datetime | None = None


class PendingTask(BaseModel):
    """A pending action item for the warehouse dashboard."""
    task_type: str         # 'staged_item', 'audit', 'spot_check'
    title: str             # "Staged: 5 items for Truck #2"
    subtitle: str | None = None  # "2 hours ago"
    severity: str = "normal"  # 'normal', 'warning', 'critical'
    # Contextual IDs for navigation
    audit_id: int | None = None
    part_id: int | None = None
    stock_id: int | None = None
    destination_type: str | None = None
    destination_id: int | None = None


class DashboardData(BaseModel):
    """Combined dashboard payload (single request)."""
    kpis: DashboardKPIs
    recent_activity: list[ActivitySummary] = Field(default_factory=list)
    pending_tasks: list[PendingTask] = Field(default_factory=list)


# =================================================================
# INVENTORY GRID
# =================================================================

class WarehouseInventoryItem(BaseModel):
    """A single row in the warehouse inventory grid."""
    part_id: int
    part_code: str | None = None
    part_name: str
    category_id: int | None = None
    category_name: str | None = None
    brand_id: int | None = None
    brand_name: str | None = None
    unit_of_measure: str = "each"
    shelf_location: str | None = None
    bin_location: str | None = None
    # Stock levels
    warehouse_qty: int = 0
    pulled_qty: int = 0
    truck_qty: int = 0
    total_qty: int = 0
    # Targets
    min_stock_level: int = 0
    target_stock_level: int = 0
    max_stock_level: int = 0
    # Health
    stock_status: str = "in_range"  # low_stock, overstock, in_range, winding_down, zero
    health_pct: float = 0.0        # 0-100 for progress bar
    # Pricing (permission-gated)
    unit_cost: float | None = None
    total_value: float | None = None
    # Forecast
    forecast_days_until_low: int | None = None
    # Supplier info
    primary_supplier_name: str | None = None
    # QR tagging
    is_qr_tagged: bool = False


# =================================================================
# STAGING
# =================================================================

class StagingItem(BaseModel):
    """A single pulled/staged stock entry with destination tag."""
    stock_id: int
    part_id: int
    part_name: str
    part_code: str | None = None
    qty: int
    supplier_name: str | None = None
    # Destination tag
    destination_type: str | None = None  # 'truck' or 'job'
    destination_id: int | None = None
    destination_label: str | None = None
    # Who pulled and when
    tagged_by_name: str | None = None
    staged_at: datetime | None = None
    # Aging
    hours_staged: float = 0.0
    aging_status: str = "normal"  # 'normal', 'warning' (24h), 'critical' (48h)


class StagingGroup(BaseModel):
    """Staged items grouped by destination."""
    destination_type: str | None = None
    destination_id: int | None = None
    destination_label: str
    items: list[StagingItem] = Field(default_factory=list)
    total_qty: int = 0
    oldest_hours: float = 0.0
    aging_status: str = "normal"


# =================================================================
# AUDIT
# =================================================================

class AuditStartRequest(BaseModel):
    """Request to start a new audit session."""
    audit_type: str = Field(pattern="^(spot_check|category|rolling)$")
    location_type: str = "warehouse"
    location_id: int = 1
    category_id: int | None = None  # required for category audits
    part_ids: list[int] | None = None  # for spot checks on specific parts


class AuditCountRequest(BaseModel):
    """Request to record a count for an audit item."""
    actual_qty: int = Field(ge=0)
    result: str = Field(pattern="^(match|discrepancy|skipped)$")
    discrepancy_note: str | None = None
    photo_path: str | None = None


class AuditItemResponse(BaseModel):
    """An individual audit item (for card-swipe UI)."""
    id: int
    audit_id: int
    part_id: int
    part_name: str
    part_code: str | None = None
    shelf_location: str | None = None
    image_url: str | None = None
    expected_qty: int
    actual_qty: int | None = None
    result: str = "pending"
    discrepancy_note: str | None = None
    photo_path: str | None = None
    counted_at: str | None = None


class AuditProgress(BaseModel):
    """Progress stats for an ongoing audit."""
    total_items: int = 0
    counted: int = 0
    matched: int = 0
    discrepancies: int = 0
    skipped: int = 0
    pending: int = 0
    pct_complete: float = 0.0


class AuditResponse(BaseModel):
    """Full audit session response."""
    id: int
    audit_type: str
    location_type: str
    location_id: int
    category_id: int | None = None
    category_name: str | None = None
    status: str
    started_by: int
    started_by_name: str | None = None
    completed_at: str | None = None
    progress: AuditProgress = Field(default_factory=AuditProgress)
    notes: str | None = None
    created_at: datetime | None = None


class AuditSummary(BaseModel):
    """Summary shown after completing an audit."""
    audit_id: int
    audit_type: str
    status: str
    progress: AuditProgress
    adjustments_needed: int = 0  # discrepancies that need stock adjustment
    has_unapplied_adjustments: bool = False


# =================================================================
# SUPPLIER PREFERENCES
# =================================================================

class SupplierPreferenceResponse(BaseModel):
    """Resolved preferred supplier for a scope."""
    scope_type: str | None = None
    scope_id: int | None = None
    supplier_id: int | None = None
    supplier_name: str | None = None
    resolved_from: str | None = None  # which level it was resolved from


class SupplierPreferenceSet(BaseModel):
    """Request to set a preferred supplier for a scope level."""
    scope_type: str = Field(pattern="^(category|style|type|part)$")
    scope_id: int
    supplier_id: int


# =================================================================
# LOCATIONS HELPER
# =================================================================

class LocationOption(BaseModel):
    """An available from/to location for the wizard."""
    location_type: str
    location_id: int
    label: str       # "Warehouse", "Truck #2", "Job #1042"
    sub_label: str | None = None  # "Main", "Mike's Truck"


# =================================================================
# RECEIVE STOCK (add new stock to warehouse)
# =================================================================

class ReceiveStockItem(BaseModel):
    """A single line item when receiving stock into the warehouse."""
    part_id: int
    qty: int = Field(ge=1, description="Quantity to add")
    shelf_location: str | None = None  # General area (Row A, Shelf 3)
    bin_location: str | None = None    # Specific bin (Bin 12) — optional
    supplier_id: int | None = None     # Which supplier provided this stock
    notes: str | None = None           # Optional notes for this line


class ReceiveStockRequest(BaseModel):
    """Request to receive stock into the warehouse.

    This is stock entering the system from outside (delivery, found stock, etc.)
    — not a transfer between locations.
    """
    items: list[ReceiveStockItem] = Field(min_length=1, max_length=50)
    reason: str | None = "Stock Received"  # Overall reason
    notes: str | None = None               # Overall notes
    reference_number: str | None = None    # PO number, delivery ref, etc.


class ReceiveStockResult(BaseModel):
    """Result of receiving stock."""
    success: bool = True
    items_received: int = 0
    total_qty: int = 0
    movement_ids: list[int] = Field(default_factory=list)

File: src/lib/types/warehouse.ts
/**
 * Warehouse types — movements, rules, dashboard, inventory, receive stock, staging,
 * pending pulls, audit, supplier preferences, locations, wizard search,
 * companion rules, alternatives.
 */

// ═══════════════════════════════════════════════════════════════════
// WAREHOUSE MODULE (Phase 3)
// ═══════════════════════════════════════════════════════════════════

// ── Movement Wizard ───────────────────────────────────────────────

export type LocationType = 'warehouse' | 'pulled' | 'truck' | 'trailer' | 'job';

export interface MovementLineItem {
  part_id: number;
  qty: number;
  supplier_id?: number | null;
}

export interface MovementRequest {
  from_location_type: LocationType;
  from_location_id: number;
  to_location_type: LocationType;
  to_location_id: number;
  items: MovementLineItem[];
  reason?: string | null;
  reason_detail?: string | null;
  notes?: string | null;
  reference_number?: string | null;
  job_id?: number | null;
  photo_path?: string | null;
  scan_confirmed?: boolean;
  gps_lat?: number | null;
  gps_lng?: number | null;
  destination_type?: string | null;
  destination_id?: number | null;
  destination_label?: string | null;
}

export interface ValidationError {
  field?: string | null;
  message: string;
  part_id?: number | null;
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
  warnings: string[];
}

export interface MovementPreviewLine {
  part_id: number;
  part_name: string;
  part_code?: string | null;
  qty: number;
  supplier_id?: number | null;
  supplier_name?: string | null;
  supplier_source?: string | null;
  source_before: number;
  source_after: number;
  dest_before: number;
  dest_after: number;
  unit_cost?: number | null;
  line_value?: number | null;
}

export interface MovementPreview {
  lines: MovementPreviewLine[];
  total_qty: number;
  total_value?: number | null;
  movement_type: string;
  photo_required: boolean;
  warnings: string[];
}

export interface MovementResult {
  movement_id: number;
  part_id: number;
  part_name: string;
  qty: number;
  movement_type: string;
  from_location_type?: string | null;
  to_location_type?: string | null;
}

export interface MovementExecuteResponse {
  success: boolean;
  movements: MovementResult[];
  total_items: number;
  total_qty: number;
}

export interface MovementLogEntry {
  id: number;
  part_id: number;
  part_name: string;
  part_code?: string | null;
  qty: number;
  movement_type: string;
  from_location_type?: string | null;
  from_location_id?: number | null;
  to_location_type?: string | null;
  to_location_id?: number | null;
  performed_by?: number | null;
  performer_name?: string | null;
  reason?: string | null;
  notes?: string | null;
  reference_number?: string | null;
  photo_path?: string | null;
  unit_cost?: number | null;
  unit_sell?: number | null;
  gps_lat?: number | null;
  gps_lng?: number | null;
  created_at: string;
}

// ── Movement Rules ────────────────────────────────────────────────

export interface MovementRule {
  from: string;
  to: string;
  type: string;
  photo_required: boolean;
}

export type ReasonCategories = Record<string, string[]>;

// ── Dashboard ─────────────────────────────────────────────────────

export interface DashboardKPIs {
  stock_health_pct: number;
  total_units: number;
  warehouse_value?: number | null;
  shortfall_count: number;
  pending_task_count: number;
}

export interface ActivitySummary {
  id: number;
  summary: string;
  movement_type: string;
  performer_name?: string | null;
  created_at?: string | null;
}

export interface PendingTask {
  task_type: string;
  title: string;
  subtitle?: string | null;
  severity: string;
  audit_id?: number | null;
  part_id?: number | null;
  stock_id?: number | null;
  destination_type?: string | null;
  destination_id?: number | null;
}

export interface DashboardData {
  kpis: DashboardKPIs;
  recent_activity: ActivitySummary[];
  pending_tasks: PendingTask[];
}

// ── Inventory Grid ────────────────────────────────────────────────

export type StockStatus = 'low_stock' | 'overstock' | 'in_range' | 'winding_down' | 'zero' | 'all';

export interface WarehouseInventoryItem {
  part_id: number;
  part_code?: string | null;
  part_name: string;
  category_id?: number | null;
  category_name?: string | null;
  brand_id?: number | null;
  brand_name?: string | null;
  unit_of_measure: string;
  shelf_location?: string | null;
  bin_location?: string | null;
  warehouse_qty: number;
  pulled_qty: number;
  truck_qty: number;
  total_qty: number;
  min_stock_level: number;
  target_stock_level: number;
  max_stock_level: number;
  stock_status: StockStatus;
  health_pct: number;
  unit_cost?: number | null;
  total_value?: number | null;
  forecast_days_until_low?: number | null;
  primary_supplier_name?: string | null;
  is_qr_tagged?: boolean;
}

export interface WarehouseInventoryParams {
  search?: string;
  category_id?: number;
  brand_id?: number;
  part_id?: number;
  stock_status?: StockStatus;
  sort_by?: string;
  sort_dir?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}

// ── Receive Stock ─────────────────────────────────────────────────

export interface ReceiveStockItem {
  part_id: number;
  qty: number;
  shelf_location?: string | null;
  bin_location?: string | null;
  supplier_id?: number | null;
  notes?: string | null;
}

export interface ReceiveStockRequest {
  items: ReceiveStockItem[];
  warehouse_location_id?: number;
  reason?: string | null;
  notes?: string | null;
  reference_number?: string | null;
}

export interface ReceiveStockResult {
  success: boolean;
  items_received: number;
  total_qty: number;
  movement_ids: number[];
}

// ── Staging ───────────────────────────────────────────────────────

export type AgingStatus = 'normal' | 'warning' | 'critical';

export interface StagingItem {
  stock_id: number;
  part_id: number;
  part_name: string;
  part_code?: string | null;
  qty: number;
  supplier_name?: string | null;
  destination_type?: string | null;
  destination_id?: number | null;
  destination_label?: string | null;
  tagged_by_name?: string | null;
  staged_at?: string | null;
  hours_staged: number;
  aging_status: AgingStatus;
}

export interface StagingGroup {
  destination_type?: string | null;
  destination_id?: number | null;
  destination_label: string;
  items: StagingItem[];
  total_qty: number;
  oldest_hours: number;
  aging_status: AgingStatus;
}

// ── Pending Pulls (JPO lines received but not yet staged) ─────────

export interface PendingPullItem {
  jpo_line_id: number;
  jpo_id: number;
  jpo_number: string;
  job_id: number | null;
  job_name: string | null;
  job_number: string | null;
  part_id: number;
  part_name: string;
  part_code: string | null;
  qty_received: number;
  qty_already_pulled: number;
  qty_pending: number;
  warehouse_qty: number;
  priority: string;
  supplier_name: string | null;
  requested_by_name: string | null;
}

export interface PendingPullGroup {
  job_id: number | null;
  job_name: string | null;
  job_number: string | null;
  items: PendingPullItem[];
  total_pending: number;
  total_already_pulled: number;
}

// ── Audit ─────────────────────────────────────────────────────────

export type AuditType = 'spot_check' | 'category' | 'rolling';
export type AuditStatus = 'in_progress' | 'paused' | 'completed' | 'cancelled';
export type AuditItemResult = 'pending' | 'match' | 'discrepancy' | 'skipped';

export interface AuditStartRequest {
  audit_type: AuditType;
  location_type?: LocationType;
  location_id?: number;
  category_id?: number | null;
  part_ids?: number[] | null;
}

export interface AuditCountRequest {
  actual_qty: number;
  result: AuditItemResult;
  discrepancy_note?: string | null;
  photo_path?: string | null;
}

export interface AuditItemResponse {
  id: number;
  audit_id: number;
  part_id: number;
  part_name: string;
  part_code?: string | null;
  shelf_location?: string | null;
  image_url?: string | null;
  expected_qty: number;
  actual_qty?: number | null;
  result: AuditItemResult;
  discrepancy_note?: string | null;
  photo_path?: string | null;
  counted_at?: string | null;
}

export interface AuditProgress {
  total_items: number;
  counted: number;
  matched: number;
  discrepancies: number;
  skipped: number;
  pending: number;
  pct_complete: number;
}

export interface AuditResponse {
  id: number;
  audit_type: AuditType;
  location_type: LocationType;
  location_id: number;
  category_id?: number | null;
  category_name?: string | null;
  status: AuditStatus;
  started_by: number;
  started_by_name?: string | null;
  completed_at?: string | null;
  progress: AuditProgress;
  notes?: string | null;
  created_at?: string | null;
}

export interface AuditSummary {
  audit_id: number;
  audit_type: AuditType;
  status: AuditStatus;
  progress: AuditProgress;
  adjustments_needed: number;
  has_unapplied_adjustments: boolean;
}

export interface SuggestedRollingPart {
  id: number;
  name: string;
  code?: string | null;
  shelf_location?: string | null;
  category_name?: string | null;
  last_counted_at: string;
  warehouse_qty: number;
}

// ── Supplier Preferences ──────────────────────────────────────────

export type SupplierPrefScope = 'category' | 'style' | 'type' | 'part';

export interface SupplierPreferenceResponse {
  scope_type?: string | null;
  scope_id?: number | null;
  supplier_id?: number | null;
  supplier_name?: string | null;
  resolved_from?: string | null;
}

export interface SupplierPreferenceSet {
  scope_type: SupplierPrefScope;
  scope_id: number;
  supplier_id: number;
}

// ── Locations Helper ──────────────────────────────────────────────

export interface LocationOption {
  location_type: LocationType;
  location_id: number;
  label: string;
  sub_label?: string | null;
}

// ── Wizard Part Search ────────────────────────────────────────────

export interface WizardPartSearchResult {
  part_id: number;
  part_name: string;
  part_code?: string | null;
  image_url?: string | null;
  category_name?: string | null;
  shelf_location?: string | null;
  available_qty: number;
  supplier_name?: string | null;
  supplier_id?: number | null;
}


// =================================================================
// COMPANION RULES — category-level linking
// =================================================================

export interface CompanionRuleSource {
  id: number;
  category_id: number;
  category_name?: string | null;
  style_id?: number | null;
  style_name?: string | null;
}

export interface CompanionRuleTarget {
  id: number;
  category_id: number;
  category_name?: string | null;
  style_id?: number | null;
  style_name?: string | null;
}

export interface CompanionRule {
  id: number;
  name: string;
  description?: string | null;
  style_match: 'auto' | 'any' | 'explicit';
  qty_mode: 'sum' | 'max' | 'ratio';
  qty_ratio: number;
  is_active: boolean;
  sources: CompanionRuleSource[];
  targets: CompanionRuleTarget[];
  created_by?: number | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface CompanionRuleSourceCreate {
  category_id: number;
  style_id?: number | null;
}

export interface CompanionRuleTargetCreate {
  category_id: number;
  style_id?: number | null;
}

export interface CompanionRuleCreate {
  name: string;
  description?: string | null;
  style_match?: 'auto' | 'any' | 'explicit';
  qty_mode?: 'sum' | 'max' | 'ratio';
  qty_ratio?: number;
  is_active?: boolean;
  sources: CompanionRuleSourceCreate[];
  targets: CompanionRuleTargetCreate[];
}

export interface CompanionRuleUpdate {
  name?: string | null;
  description?: string | null;
  style_match?: 'auto' | 'any' | 'explicit' | null;
  qty_mode?: 'sum' | 'max' | 'ratio' | null;
  qty_ratio?: number | null;
  is_active?: boolean | null;
  sources?: CompanionRuleSourceCreate[] | null;
  targets?: CompanionRuleTargetCreate[] | null;
}


// =================================================================
// COMPANION SUGGESTIONS
// =================================================================

export interface SuggestionSource {
  id: number;
  category_id: number;
  category_name?: string | null;
  style_id?: number | null;
  style_name?: string | null;
  qty: number;
}

export interface CompanionSuggestion {
  id: number;
  rule_id?: number | null;
  target_category_id: number;
  target_style_id?: number | null;
  target_description: string;
  suggested_qty: number;
  approved_qty?: number | null;
  reason_type: 'rule' | 'learned' | 'mixed';
  reason_text: string;
  status: 'pending' | 'approved' | 'discarded';
  sources: SuggestionSource[];
  triggered_by?: number | null;
  decided_by?: number | null;
  decided_at?: string | null;
  order_id?: number | null;
  notes?: string | null;
  created_at?: string | null;
}

export interface SuggestionDecision {
  action: 'approved' | 'discarded';
  approved_qty?: number | null;
  notes?: string | null;
}


// =================================================================
// MANUAL TRIGGER — "What should I also order?"
// =================================================================

export interface ManualTriggerItem {
  category_id: number;
  style_id?: number | null;
  qty: number;
}

export interface ManualTriggerRequest {
  items: ManualTriggerItem[];
}


// =================================================================
// CO-OCCURRENCE
// =================================================================

export interface CoOccurrencePair {
  id: number;
  category_a_id: number;
  category_a_name?: string | null;
  category_b_id: number;
  category_b_name?: string | null;
  co_occurrence_count: number;
  total_jobs_a: number;
  total_jobs_b: number;
  avg_ratio_a_to_b: number;
  confidence: number;
  last_computed?: string | null;
}


// =================================================================
// COMPANION STATS
// =================================================================

export interface CompanionStats {
  total_rules: number;
  active_rules: number;
  pending_suggestions: number;
  approved_count: number;
  discarded_count: number;
  co_occurrence_pairs: number;
}


// =================================================================
// PART ALTERNATIVES — individual part cross-linking
// =================================================================

export type AlternativeRelationship = 'substitute' | 'upgrade' | 'compatible';

export interface PartAlternative {
  id: number;
  part_id: number;
  part_name?: string | null;
  part_code?: string | null;
  alternative_part_id: number;
  alternative_name?: string | null;
  alternative_code?: string | null;
  alternative_brand_name?: string | null;
  relationship: AlternativeRelationship;
  preference: number;
  notes?: string | null;
  created_by?: number | null;
  created_at?: string | null;
}

export interface PartAlternativeCreate {
  alternative_part_id: number;
  relationship?: AlternativeRelationship;
  preference?: number;
  notes?: string | null;
}

export interface PartAlternativeUpdate {
  relationship?: AlternativeRelationship | null;
  preference?: number | null;
  notes?: string | null;
}
